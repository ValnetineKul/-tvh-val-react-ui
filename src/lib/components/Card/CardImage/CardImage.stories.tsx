import React from 'react';
import { createMeta } from '../../../story-utils';
import filterImage from '../mocked-assets/filter.jpg';
import CardImage from './CardImage';

const Template = (args: React.ComponentProps<typeof CardImage>) => {
  return (
    <div style={{ width: 400, height: 400 }}>
      <CardImage {...args} />
    </div>
  );
};

export const Primitive = Template.bind({});

Primitive.args = {
  image: filterImage,
  direction: 'vertical',
};

Primitive.storyName = 'CardImage';

export default createMeta({
  component: CardImage,
  title: 'Components/Cards/CardImage',
  argTypes: {
    image: {
      control: {
        disable: true,
      },
      table: {
        type: { detail: 'Image source' },
      },
    },
    direction: {
      control: {
        disable: true,
      },
      table: {
        type: { detail: 'Used by developer to align aspect ratio for horizontal and vertical cards' },
      },
    },
  },
});
