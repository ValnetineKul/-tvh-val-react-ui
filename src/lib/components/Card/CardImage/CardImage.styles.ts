import { makeStyles } from '../../../../themes/core';

const useStyles = makeStyles({ name: 'CardImage' })(() => ({
  directionHorizontal: {
    width: 0,
    paddingLeft: '43.75%',
  },
  directionVertical: {
    height: 0,
    paddingTop: '56.25%',
  },
  sizeContain: {
    backgroundSize: 'contain',
  },
  sizeCover: {
    backgroundSize: 'cover',
  },
}));

export default useStyles;
