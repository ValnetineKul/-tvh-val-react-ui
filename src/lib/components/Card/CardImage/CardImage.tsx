import CardMedia from '@mui/material/CardMedia';

import type { FC } from 'react';
import React from 'react';
import { capitalize } from '@mui/material/utils';
import useStyles from './CardImage.styles';

export interface CardImageProps {
  image: string;
  direction: 'horizontal' | 'vertical';
  imgSize?: 'contain' | 'cover';
}

const CardImage: FC<CardImageProps> = ({ image, imgSize = 'contain', direction }) => {
  const { classes, cx } = useStyles();
  const sizeClassName = (classes as Record<string, string>)[`size${capitalize(imgSize)}`];
  const directionClassName = (classes as Record<string, string>)[`direction${capitalize(direction)}`];
  return <CardMedia className={cx(sizeClassName, directionClassName)} image={image} />;
};

export default CardImage;
