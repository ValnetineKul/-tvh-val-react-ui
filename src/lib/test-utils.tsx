import userEvent from '@testing-library/user-event';
import { fireEvent, render } from '@testing-library/react';
import { renderHook } from '@testing-library/react-hooks';
import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { theme } from '../themes/tvh';
import { createTheme, ThemeProvider, SurfaceProvider } from '../themes/core';

export const muiTheme = createTheme(theme);

const AllTheProviders: React.FC = ({ children }) => {
  return (
    <MemoryRouter>
      <ThemeProvider theme={muiTheme}>
        <SurfaceProvider color="100">{children}</SurfaceProvider>
      </ThemeProvider>
    </MemoryRouter>
  );
};

const customRender = (component: React.ReactElement, options?: Parameters<typeof render>[1]) =>
  render(component, { wrapper: AllTheProviders, ...options });

const customRenderHook: typeof renderHook = (hook, options) =>
  renderHook(hook, { wrapper: AllTheProviders, ...options });

export const initResponsiveTest = (size: number | 'Mobile' | 'Tablet' | 'Desktop' | 'LgDesktop' | 'XlDesktop') => {
  let width: number;
  if (size === 'Mobile') width = muiTheme.breakpoints.values.sm - 1;
  else if (size === 'Tablet') width = muiTheme.breakpoints.values.sm;
  else if (size === 'Desktop') width = muiTheme.breakpoints.values.md;
  else if (size === 'LgDesktop') width = muiTheme.breakpoints.values.lg;
  else if (size === 'XlDesktop') width = muiTheme.breakpoints.values.xl;
  else width = size;

  Object.defineProperty(window, 'innerWidth', {
    writable: true,
    configurable: true,
    value: width,
  });

  window.matchMedia = jest.fn().mockImplementation((query) => {
    const queryWidthInterval: string[] = query.match(/\d+(\.\d+)?/g);

    let matches;
    if (queryWidthInterval.length === 2) {
      matches = width >= +queryWidthInterval[0] && width < +queryWidthInterval[1];
    } else {
      const queryWidth = +queryWidthInterval[0];
      const isDown = /max/.test(query);

      matches = isDown ? width <= queryWidth : width >= queryWidth;
    }

    return {
      matches,
      media: query,
      onchange: null,
      addListener: jest.fn(),
      removeListener: jest.fn(),
    };
  });
};

const customUserEvent = Object.assign(userEvent, {
  mouseDown: fireEvent.mouseDown,
  mouseUp: fireEvent.mouseUp,
  mouseMove: fireEvent.mouseMove,
  scroll: fireEvent.scroll,
  error: fireEvent.error,
});

export { screen, within, waitFor, act, waitForElementToBeRemoved } from '@testing-library/react';
// override render method
export { customRender as render, customRenderHook as renderHook, customUserEvent as userEvent };
