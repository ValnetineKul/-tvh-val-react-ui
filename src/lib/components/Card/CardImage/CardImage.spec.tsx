import React from 'react';
import { render } from '../../../test-utils';
import CardImage from './CardImage';
import filterImage from '../mocked-assets/filter.jpg';

describe('CardImage', () => {
  it('Should render correctly', () => {
    const { container } = render(<CardImage image={filterImage} direction="vertical" imgSize="contain" />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
