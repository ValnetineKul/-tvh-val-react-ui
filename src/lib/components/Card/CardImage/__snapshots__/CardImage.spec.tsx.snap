// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardImage Should render correctly 1`] = `
<div
  class="MuiCardMedia-root tss-1xjno9i-CardImage-sizeContain-CardImage-directionVertical mui-16cvb1l-MuiCardMedia-root"
  role="img"
  style="background-image: url(filter.jpg);"
/>
`;
